components/resource_control/src/resource_group.rs

```diff
@@ -0,0 +1,480 @@
+        self.resource_groups.get(&name.to_ascii_lowercase())
+    }
+
+    pub fn get_all_resource_groups(&self) -> Vec<ResourceGroup> {
+        self.resource_groups.iter().map(|g| g.clone()).collect()
+    }
+
+    pub fn derive_controller(&self, name: String, is_read: bool) -> Arc<ResourceController> {
+        let controller = Arc::new(ResourceController::new(name, is_read));
+        self.registry.lock().unwrap().push(controller.clone());
+        for g in &self.resource_groups {
+            let ru_quota = Self::get_ru_setting(g.value(), controller.is_read);
+            controller.add_resource_group(g.key().clone().into_bytes(), ru_quota);
+        }
+
+        controller
+    }
+
+    pub fn advance_min_virtual_time(&self) {
+        for controller in self.registry.lock().unwrap().iter() {
+            controller.update_min_virtual_time();
+        }
+    }
+}
+
+pub struct ResourceController {
+    // resource controller name is not used currently.
+    #[allow(dead_code)]
+    name: String,
+    // We handle the priority differently between read and write request:
+    // 1. the priority factor is calculate based on read/write RU settings.
+    // 2. for read request, we increase a constant virtual time delta at each `get_priority` call
+    //    because the cost can't be calculated at start, so we only increase a constant delta and
+    //    increase the real cost after task is executed; but don't increase it at write because
+    //    the cost is known so we just pre-consume it.
+    is_read: bool,
+    // Track the maximum ru quota used to calculate the factor of each resource group.
+    // factor = max_ru_quota / group_ru_quota * 10.0
+    max_ru_quota: Mutex<u64>,
+    // record consumption of each resource group, name --> resource_group
+    resource_consumptions: DashMap<Vec<u8>, GroupPriorityTracker>,
+
+    last_min_vt: AtomicU64,
+}
+
+impl ResourceController {
+    pub fn new(name: String, is_read: bool) -> Self {
+        let controller = Self {
+            name,
+            is_read,
+            max_ru_quota: Mutex::new(DEFAULT_MAX_RU_QUOTA),
+            resource_consumptions: DashMap::new(),
+            last_min_vt: AtomicU64::new(0),
+        };
+        // add the "default" resource group
+        controller.add_resource_group(DEFAULT_RESOURCE_GROUP_NAME.as_bytes().to_owned(), 0);
+        controller
+    }
+
+    fn calculate_factor(max_quota: u64, quota: u64) -> u64 {
+        if quota > 0 {
+            // we use max_quota / quota as the resource group factor, but because we need to
+            // cast the value to integer, so we times it by 10 to ensure the accuracy is
+            // enough.
+            (max_quota as f64 / quota as f64 * 10.0).round() as u64
+        } else {
+            1
+        }
+    }
+
+    fn add_resource_group(&self, name: Vec<u8>, ru_quota: u64) {
+        let mut max_ru_quota = self.max_ru_quota.lock().unwrap();
```

Comment:
why not use atomic

Reply:
Because when triggering `adjust_all_resource_group_factors` we need to ensure the max ru quota is not changed. Since we don't assume that the CRUD of resource group happens very often, use a mutex can simple the logic.
If we use atomic here, when there are too new resource groups calling add at the same time and each of them triggers adjust all factors, we can ensure the later (which is large) takes effect for all groups.

Reply:
Added a comment to explain this mutex.